### ISSUE TITLE
[logrotate|dhparam] Both still force the installation of cron which isn't necessary

### ISSUE THREAD

**Issue by Alphix** [debops/debops#1990]
@drybjed : PR #1988 and #1989 sought to make the dependency on ``cron`` softer by relying on the ``debops.cron`` role (which can then install ``systemd-cron``). I'd been working on something similar, so good job 👍🏻 

However, both ``logrotate`` and ``dhparam`` include native systemd timers, so they don't actually need any cron (including ``systemd-cron``) on systems which use ``systemd``.

The ``logrotate`` package already expresses this in [its dependencies](https://packages.debian.org/bullseye/logrotate) (``cron | anacron | cron-daemon | systemd-sysv``), so that should be automatic (i.e. no dependency on ``debops.cron`` is strictly necessary?). ``debops.dhparam`` is a little bit trickier, but I think the ``debops.cron`` role dependency should be made conditional on not having ``systemd``?

**Comment by drybjed** [comment]
Ansible doesn't really like having conditions on the playbook level, it might make things weird in certain cases, for example with `flush_handlers` task. I think the proper way to deal with this would be to add support for `systemd-cron` in the `debops.cron` role. That way we can configure everything that's needed for it in one place, then perhaps add conditions in `dhparam` and `logrotate` roles to only activate `cron`-related options when the native `cron` is present, via `debops.cron` Ansible fact.

**Comment by Alphix** [comment]
That could be, I haven't tested setting any conditionals on the roles in the playbooks.

But I think it should ultimately also be possible to do completely without `systemd-cron` in a system which only has native timers...dunno how realistic that is ATM.

**Comment by Alphix** [comment]
``logrotate`` doesn't even use ``systemd-cron`` in a ``systemd`` environment....so is your plan that `debops.cron` should be changed so that it can be informed that native timers exist and do a no-op then?

**Comment by drybjed** [comment]
Sure, it can be option there as well. `cron`, `systemd-cron`, `none`.

**Comment by ser** [comment]
```systemd-cron``` in nothing else than systemd timers packaged with handy crontab script. I can't imagine making it simpler to be honest.

**Comment by Alphix** [comment]
> `systemd-cron` in nothing else than systemd timers packaged with handy crontab script

Yeah, I know....but the ``logrotate`` and ``dhparams`` roles now express a role dependency on ``debops.cron`` (so that `systemd-cron` can be used...ostensibly....but `systemd-cron` isn't _actually_ used by either of them)

**Comment by ser** [comment]
but you have a tool (crontab script) which can modify settings locally in a friendly and well documented way. IMHO it's much simpler than directly editing systemd timers.

**Comment by Alphix** [comment]
> but you have a tool (crontab script) which can modify settings locally in a friendly and well documented way. IMHO it's much simpler than directly editing systemd timers.

But unless I'm mistaken it won't help you at all with the native timers of `logrotate` and `dhparams` (which is what'll be executed, not the additional `cron` jobs that they include)

**Comment by ser** [comment]
please notice syntax of these files, they are crontab-like. to swap them into systemd native files we would need keep two copies of them, one in crontab syntax and one with systemd timer syntax, which would make debops maintaner's life harder.

**Comment by Alphix** [comment]
> please notice syntax of these files, they are crontab-like. to swap them into systemd native files we would need keep two copies of them, one in crontab syntax and one with systemd timer syntax, which would make debops maintaner harder.

There are already two (well, three) files in [logrotate](https://packages.debian.org/bullseye/amd64/logrotate/filelist) and in [dhparam](https://github.com/debops/debops/blob/master/ansible/roles/dhparam/tasks/main.yml), that's not really what this is about...it's about installing `systemd-cron` when it won't be used

**Comment by ser** [comment]
OK, you are correct, I checked content of previous, older version of logrotate.

**Comment by ser** [comment]
btw i have joined this conversation as i am trying to de-crond my new containers - and this small change made by drybjed at least eliminated crond process which is at least half way forward :)

**Comment by Alphix** [comment]
> btw i have joined this conversation as i am trying to de-crond my new containers - and this small change made by drybjed at least eliminated crond process which is at least half way forward :)

I agree....I managed to step on this exact same landmine just a few days ago for the exact same reasons, I wanted to de-cron my VMs, so I set ``cron__base_packages`` to ``systemd-cron`` and the end result was an eternal back-and-forth between `cron` (being installed e.g. by `debops.logrotate`) and `systemd-cron` (installed by `debops.cron`).

Don't get me wrong. It's better now, but it could be even better :)

**Comment by ser** [comment]
it's funny that we started to do the same in the same time :)

**Comment by Alphix** [comment]
@drybjed : ok, I've given this some more thought, and I have a suggestion/idea.

We could add a ``cron__dependent_config`` variable to the ``cron`` role, where other roles that use ``cron`` as a dependent role could specify the systemd timers and/or cron jobs they want to have executed (with cron jobs and timers being alternatives, so it's an *or*, not *and*).

The ``cron`` role can then do the right thing (since systemd timers and cron jobs would be expected to be alternatives, just like it is now if you install a timer by the same name as an existing cron job). So, if you include the `cron` role with a ``cron__dependent_config`` containing a system timer and a cron job, *and* the host is running `systemd`, that would pretty much be a noop. If you call it only with a cron job on a system running `systemd`, it'd make sure some kind of cron daemon is installed, etc (and if you only call it with a systemd timer on a system without systemd, it could error out).

This would also make the role dependencies much clearer. Right now, if you take a random role, say `owncloud-nginx`, then its playbook reads, in part (taken from `ansible/playbooks/service/owncloud-nginx.yml`):

```
    - role: cron
      tags: [ 'role::cron', 'skip::cron' ]

    - role: logrotate
      tags: [ 'role::logrotate', 'skip::logrotate' ]
      logrotate__dependent_config:
        - '{{ php__logrotate__dependent_config }}'
        - '{{ owncloud__logrotate__dependent_config }}'
```

And instead it'd read:

```
    - role: cron
      tags: [ 'role::cron', 'skip::cron' ]
      cron__dependent_config:
        - '{{ logrotate__cron__dependent_config }}'

    - role: logrotate
      tags: [ 'role::logrotate', 'skip::logrotate' ]
      logrotate__dependent_config:
        - '{{ php__logrotate__dependent_config }}'
        - '{{ owncloud__logrotate__dependent_config }}'
```

Where `logrotate__cron__dependent_config` would be something like:

```
logrotate__cron__dependent_config:
  - cron_job: '/etc/blablabla'
```

While, say `dhparam__cron__dependent_config` would be something like:

```
dhparam__cron__dependent_config:
  - cron_job: '/etc/foo...'
    systemd_timer: '/etc/bar...'
```

Which has the added benefit that it'd make it much clearer why e.g. `owncloud-nginx` has a dependency on `cron` (i.e. its due to the `logrotate` role). In addition, roles that have native timers wouldn't cause the installation of a `cron` daemon at all (if the host is using systemd).

What do you think?

**Comment by drybjed** [comment]
The idea of delegating management of timers to the `cron` role looks nice. I'm not sure if I understood you correctly, but it seems like you expect the `cron` role to install `cron` daemon if any cron jobs are configured via dependent variables. I think that would be counterintuitive - if the admin doesn't want `cron` (hi @ser), then I don't see the point of pushing it on him/her via some random role dependency.

I would see it like this: the `cron` role has two lists, `cron__dependent_cron_jobs` and `cron__dependent_timers`. Other roles can define both as they see fit, using whatever syntax we devise for this. In the `cron` role, we can decide which mode of operation we use on a given host - `systemd` timers, `cron` jobs, maybe both because of reasons. And then, the role can configure either one as needed.

With this scheme, not all roles will initially provide both cron jobs and timers, but this will be a good incentive to add missing configuration. We can support `systemd` and `non-systemd` environments easily this way. And the local administrator has complete control over what technology is used.

Edit: depending on how big the timers support is, I think that we could put it in a separate role, say `timerd` (`systemd_timerd` seems a bit long, and we have `journald` as a precedence). Then disabling the whole `cron` role would leave us with only timers and vice versa.

**Comment by Alphix** [comment]
> The idea of delegating management of timers to the `cron` role looks nice. I'm not sure if I understood you correctly, but it seems like you expect the `cron` role to install `cron` daemon if any cron jobs are configured via dependent variables. I think that would be counterintuitive - if the admin doesn't want `cron` (hi @ser), then I don't see the point of pushing it on him/her via some random role dependency.

Well yeah, the idea was that a given role could say (via dependency variables), "here's a cron script and/or systemd timer...just make sure *one* of them will be installed/executed, I'll leave the details up to you" ("you" being the `cron` role).

Then the `cron` role would have enough information to do the right thing....which would mean doing nothing for all the roles that provide timers if the system is already running `systemd`.

And I think @ser is fine with `systemd-cron` (as the fallback for roles that only have cron jobs). I think the priority should be to make sure that the cron job or systemd timer is executed. It'd be quite surprising to install, say, `dhparams` and realise that no params are (re)generated because I fiddled around with the configuration for the `cron` role. Over time we can try to make sure all roles have native timers for their tasks...

Anyway, I think I'll write a draft PR, it's easier to explain things in patches than in words :)
